src/lib/comms/sol-lwm2m.c

```diff
@@ -2822,19 +2822,25 @@ handle_read(struct sol_lwm2m_client *client,
         }
     }
 
-    SOL_VECTOR_FOREACH_IDX (&resources, res, i) {
-        r = setup_tlv(res, &buf);
-        SOL_INT_CHECK_GOTO(r, < 0, err_exit);
-        sol_lwm2m_resource_clear(res);
-    }
-
     r = add_coap_int_option(resp, SOL_COAP_OPTION_CONTENT_FORMAT,
         &format, sizeof(format));
     SOL_INT_CHECK_GOTO(r, < 0, err_exit);
 
     r = set_packet_payload(resp, (const uint8_t *)buf.data, buf.used);
     SOL_INT_CHECK_GOTO(r, < 0, err_exit);
 
+    SOL_VECTOR_FOREACH_IDX (&resources, res, i) {
+        r = setup_tlv(res, &buf);
+        SOL_INT_CHECK_GOTO(r, < 0, err_exit);
+    }
+
+    /* Go through the list another time to prevent double free in case of
+     * an error in setup_tlv (which triggers free for all resources in the
+     * error path processing). */
+    SOL_VECTOR_FOREACH_IDX (&resources, res, i) {
```

Comment:
but error in `setup_tlv()` is `goto err_exit`, it would not walk the list once more. I don't get what you mean here :-(